src/main/java/com/amazon/opendistroforelasticsearch/sql/parser/HavingParser.java

```diff
@@ -0,0 +1,118 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.parser;
+
+import com.alibaba.druid.sql.ast.SQLExpr;
+import com.alibaba.druid.sql.ast.SQLObject;
+import com.alibaba.druid.sql.ast.expr.SQLAggregateExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBetweenExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOpExpr;
+import com.alibaba.druid.sql.ast.expr.SQLIdentifierExpr;
+import com.alibaba.druid.sql.ast.expr.SQLInListExpr;
+import com.alibaba.druid.sql.ast.expr.SQLNotExpr;
+import com.alibaba.druid.sql.dialect.mysql.visitor.MySqlASTVisitorAdapter;
+import com.amazon.opendistroforelasticsearch.sql.domain.Field;
+import com.amazon.opendistroforelasticsearch.sql.domain.Where;
+import com.amazon.opendistroforelasticsearch.sql.exception.SqlParseException;
+
+import java.util.List;
+
+/**
+ * Parse expression in the Having clause.
+ */
+public class HavingParser {
+    private final WhereParser whereParser;
+    private final List<Field> havingFields;
+    private final HavingConditionRewriter havingConditionRewriter;
+
+    public HavingParser(WhereParser whereParser, List<Field> havingFields) {
+        this.whereParser = whereParser;
+        this.havingFields = havingFields;
+        this.havingConditionRewriter = new HavingConditionRewriter();
+    }
+
+    public void parseWhere(SQLExpr expr, Where where) throws SqlParseException {
+        expr.accept(havingConditionRewriter);
```

Comment:
How about we separate the rewrite and parseWhere? I'm thinking in this case this class doesn't need to share the `havingFields` with outside, for example:

```
    public List<Field> rewriteHavingClause() {
        // create rewriter and rewrite
        return Collections.unmodifiableList(rewriter.havingFields);
    }

    private static class HavingConditionRewriter ... {
        List<Field> havingFields = new ArrayList<>();
    }

Reply:
The idea to implement HavingParser is to encapsulate the procedure and differentate with WhereParser.